File: obsidian_excel_automation/generators/excel_generator.py
"""
Base Excel generator - creates and manages Excel workbooks
"""

from openpyxl import Workbook
from openpyxl.worksheet.worksheet import Worksheet
from pathlib import Path
from typing import List, Dict, Any, Optional

from .style_manager import StyleManager
from ..config import Config
from ..utils.logger import logger


class ExcelGenerator:
    """Base class for generating Excel workbooks"""

    def __init__(self, config: Config):
        """
        Initialize Excel generator

        Args:
            config: Application configuration
        """
        self.config = config
        self.style_manager = StyleManager(config.styling)
        self.workbook: Optional[Workbook] = None

    def create_workbook(self) -> Workbook:
        """Create a new Excel workbook"""
        logger.info("Creating new Excel workbook...")
        self.workbook = Workbook()

        # Remove default sheet
        if 'Sheet' in self.workbook.sheetnames:
            self.workbook.remove(self.workbook['Sheet'])

        return self.workbook

    def add_sheet(self, title: str, index: Optional[int] = None) -> Worksheet:
        """
        Add a new sheet to the workbook

        Args:
            title: Sheet title
            index: Position (None = append at end)

        Returns:
            Created worksheet
        """
        if not self.workbook:
            self.create_workbook()

        logger.debug(f"Adding sheet: {title}")

        if index is not None:
            ws = self.workbook.create_sheet(title, index)
        else:
            ws = self.workbook.create_sheet(title)

        return ws

    def save(self, output_path: Path):
        """
        Save workbook to file

        Args:
            output_path: Output file path
        """
        if not self.workbook:
            raise ValueError("No workbook to save")

        # Ensure output directory exists
        output_path.parent.mkdir(parents=True, exist_ok=True)

        logger.info(f"Saving workbook to: {output_path}")
        self.workbook.save(output_path)
        logger.info("✅ Workbook saved successfully")

    def add_summary_section(
        self,
        ws: Worksheet,
        title: str,
        data: Dict[str, Any],
        start_row: int,
        start_col: int = 1
    ) -> int:
        """
        Add a summary section with key-value pairs

        Args:
            ws: Worksheet
            title: Section title
            data: Dictionary of key-value pairs
            start_row: Starting row
            start_col: Starting column

        Returns:
            Next available row
        """
        current_row = start_row

        # Title
        cell = ws.cell(row=current_row, column=start_col)
        cell.value = title
        self.style_manager.apply_subheader_style(cell)
        ws.merge_cells(
            start_row=current_row,
            start_column=start_col,
            end_row=current_row,
            end_column=start_col + 1
        )
        current_row += 1

        # Data rows
        for key, value in data.items():
            # Key
            key_cell = ws.cell(row=current_row, column=start_col)
            self.style_manager.apply_normal_style(key_cell, key, bold=True)

            # Value
            value_cell = ws.cell(row=current_row, column=start_col + 1)
            self.style_manager.apply_normal_style(value_cell, str(value))

            current_row += 1

        return current_row + 1  # Add spacing

    def add_table(
        self,
        ws: Worksheet,
        headers: List[str],
        data: List[List[Any]],
        start_row: int,
        start_col: int = 1,
        apply_alternating_colors: bool = True
    ) -> int:
        """
        Add a table with headers and data

        Args:
            ws: Worksheet
            headers: Column headers
            data: Data rows
            start_row: Starting row
            start_col: Starting column
            apply_alternating_colors: Apply alternating row colors

        Returns:
            Next available row
        """
        current_row = start_row

        # Headers
        self.style_manager.format_header_row(ws, current_row, headers, start_col)
        current_row += 1

        # Data
        for row_data in data:
            self.style_manager.format_data_row(ws, current_row, row_data, start_col)
            current_row += 1

        # Apply table styling
        if apply_alternating_colors and data:
            end_row = start_row + len(data)
            end_col = start_col + len(headers) - 1
            self.style_manager.apply_table_style(
                ws, start_row + 1, end_row, start_col, end_col
            )

        return current_row + 1  # Add spacing

    def add_metric_boxes(
        self,
        ws: Worksheet,
        metrics: List[Dict[str, Any]],
        start_row: int,
        start_col: int = 1,
        boxes_per_row: int = 4
    ) -> int:
        """
        Add metric boxes in a grid layout

        Args:
            ws: Worksheet
            metrics: List of {title, value, target} dicts
            start_row: Starting row
            start_col: Starting column
            boxes_per_row: Number of boxes per row

        Returns:
            Next available row
        """
        current_row = start_row
        current_col = start_col

        for idx, metric in enumerate(metrics):
            # Box title
            title_cell = ws.cell(row=current_row, column=current_col)
            title_cell.value = metric.get('title', '')
            self.style_manager.apply_subheader_style(title_cell)

            # Merge cells for box
            ws.merge_cells(
                start_row=current_row,
                start_column=current_col,
                end_row=current_row,
                end_column=current_col + 1
            )

            # Value row
            value_row = current_row + 1

            # Current value
            value_cell = ws.cell(row=value_row, column=current_col)
            value_cell.value = metric.get('value', '')
            self.style_manager.apply_normal_style(value_cell, bold=True, centered=True)

            # Target value (if provided)
            if 'target' in metric:
                target_cell = ws.cell(row=value_row, column=current_col + 1)
                target_cell.value = f"목표: {metric['target']}"
                self.style_manager.apply_normal_style(target_cell, centered=True)

            # Progress row (if provided)
            if 'progress' in metric:
                progress_row = current_row + 2
                progress_cell = ws.cell(row=progress_row, column=current_col)
                progress_cell.value = metric['progress']
                self.style_manager.apply_normal_style(progress_cell, centered=True)
                ws.merge_cells(
                    start_row=progress_row,
                    start_column=current_col,
                    end_row=progress_row,
                    end_column=current_col + 1
                )

            # Move to next position
            if (idx + 1) % boxes_per_row == 0:
                # New row
                current_row += 4  # Box height (title + value + progress + spacing)
                current_col = start_col
            else:
                # Next column
                current_col += 3  # Box width + spacing

        return current_row + 4 if current_col > start_col else current_row

    def set_sheet_properties(
        self,
        ws: Worksheet,
        freeze_panes: bool = True,
        auto_filter: bool = True,
        column_widths: Optional[Dict[int, int]] = None
    ):
        """
        Set common sheet properties

        Args:
            ws: Worksheet
            freeze_panes: Freeze top row
            auto_filter: Enable auto filter on headers
            column_widths: Custom column widths
        """
        if freeze_panes:
            self.style_manager.freeze_panes(ws, row=2, col=1)

        if auto_filter and ws.max_row > 1:
            ws.auto_filter.ref = ws.dimensions

        if column_widths:
            self.style_manager.set_column_widths(ws, column_widths)
        else:
            self.style_manager.auto_adjust_column_widths(ws)

File: src/generators/ExcelGenerator.ts
/**
 * Base Excel generator class
 * Equivalent to Python generators/excel_generator.py
 * v2.0 - Enhanced with ConfigManager for i18n
 */

import ExcelJS from 'exceljs';
import { StyleManager } from './StyleManager';
import { ExcelAutomationSettings } from '../types/settings';
import { ConfigManager } from '../services/ConfigManager';
import { StyleConfig } from '../types/config';

/**
 * Base class for Excel workbook generation
 */
export class ExcelGenerator {
  protected workbook: ExcelJS.Workbook | null = null;
  protected styleManager: StyleManager;
  protected configManager?: ConfigManager;

  constructor(protected settings: ExcelAutomationSettings, configManager?: ConfigManager) {
    this.configManager = configManager;

    // Use style config from ConfigManager if available
    if (configManager) {
      const styleConfig = configManager.getStyle();
      this.styleManager = new StyleManager(this.convertStyleConfig(styleConfig));
    } else {
      this.styleManager = new StyleManager(settings.styling);
    }
  }

  /**
   * Convert v2 StyleConfig to legacy StylingOptions format
   */
  private convertStyleConfig(styleConfig: StyleConfig): ExcelAutomationSettings['styling'] {
    return {
      headerColor: styleConfig.colors.headerBackground,
      subheaderColor: styleConfig.colors.subheaderBackground,
      priorityColors: {
        P0: styleConfig.colors.priority.p0,
        P1: styleConfig.colors.priority.p1,
        P2: styleConfig.colors.priority.p2,
      },
      statusColors: styleConfig.colors.status,
    };
  }

  /**
   * Create a new workbook
   */
  createWorkbook(): ExcelJS.Workbook {
    this.workbook = new ExcelJS.Workbook();

    // Set workbook metadata
    this.workbook.creator = 'Obsidian Excel Automation';
    this.workbook.created = new Date();
    this.workbook.modified = new Date();

    return this.workbook;
  }

  /**
   * Get current workbook (creates one if not exists)
   */
  getWorkbook(): ExcelJS.Workbook {
    if (!this.workbook) {
      this.createWorkbook();
    }
    return this.workbook!;
  }

  /**
   * Add a new worksheet
   */
  addSheet(title: string, index?: number): ExcelJS.Worksheet {
    const wb = this.getWorkbook();

    const ws = wb.addWorksheet(title, {
      properties: {
        tabColor: { argb: 'FF4472C4' },
      },
    });

    // Set default row height
    ws.properties.defaultRowHeight = 20;

    return ws;
  }

  /**
   * Add a table with headers and data
   */
  addTable(
    ws: ExcelJS.Worksheet,
    headers: string[],
    data: (string | number | null | undefined)[][],
    startRow: number,
    startCol: number = 1,
    options?: {
      alternateColors?: boolean;
      applyPriorityToColumn?: number;
      applyStatusToColumn?: number;
    }
  ): number {
    // Add header row
    headers.forEach((header, idx) => {
      const cell = ws.getCell(startRow, startCol + idx);
      this.styleManager.applyHeaderStyle(cell, header);
    });

    // Add data rows
    let currentRow = startRow + 1;
    data.forEach((rowData, rowIdx) => {
      const alternate = options?.alternateColors && rowIdx % 2 === 1;

      rowData.forEach((value, colIdx) => {
        const cell = ws.getCell(currentRow, startCol + colIdx);
        const displayValue = value !== null && value !== undefined ? String(value) : '-';
        this.styleManager.applyNormalStyle(cell, displayValue);

        if (alternate) {
          this.styleManager.applyAlternateStyle(cell);
        }

        // Apply priority styling if configured
        if (options?.applyPriorityToColumn === colIdx && value) {
          this.styleManager.applyPriorityStyle(cell, String(value));
        }

        // Apply status styling if configured
        if (options?.applyStatusToColumn === colIdx && value) {
          this.styleManager.applyStatusStyle(cell, String(value));
        }
      });

      currentRow++;
    });

    return currentRow + 1;
  }

  /**
   * Add a summary section with key-value pairs
   */
  addSummarySection(
    ws: ExcelJS.Worksheet,
    title: string,
    data: Record<string, string | number>,
    startRow: number,
    startCol: number = 1
  ): number {
    // Title row
    const titleCell = ws.getCell(startRow, startCol);
    this.styleManager.applySubheaderStyle(titleCell, title);
    ws.mergeCells(startRow, startCol, startRow, startCol + 1);

    let currentRow = startRow + 1;

    // Key-value pairs
    for (const [key, value] of Object.entries(data)) {
      const keyCell = ws.getCell(currentRow, startCol);
      keyCell.value = key;
      keyCell.font = { bold: true, size: 10 };
      keyCell.border = this.getBorder();

      const valueCell = ws.getCell(currentRow, startCol + 1);
      valueCell.value = value;
      valueCell.font = { size: 10 };
      valueCell.border = this.getBorder();

      currentRow++;
    }

    return currentRow + 1;
  }

  /**
   * Add metric boxes in a grid layout
   */
  addMetricBoxes(
    ws: ExcelJS.Worksheet,
    metrics: { label: string; value: string | number }[],
    startRow: number,
    startCol: number = 1,
    boxesPerRow: number = 4
  ): number {
    let currentRow = startRow;
    let currentCol = startCol;

    metrics.forEach((metric, idx) => {
      // Label cell
      const labelCell = ws.getCell(currentRow, currentCol);
      this.styleManager.applySubheaderStyle(labelCell, metric.label);

      // Value cell (below label)
      const valueCell = ws.getCell(currentRow + 1, currentCol);
      valueCell.value = metric.value;
      valueCell.font = { bold: true, size: 14 };
      valueCell.alignment = { horizontal: 'center', vertical: 'middle' };
      valueCell.border = this.getBorder();

      currentCol++;

      // Move to next row if needed
      if ((idx + 1) % boxesPerRow === 0) {
        currentRow += 3; // Skip row for spacing
        currentCol = startCol;
      }
    });

    // Calculate final row
    const totalRows = Math.ceil(metrics.length / boxesPerRow);
    return startRow + totalRows * 3;
  }

  /**
   * Merge cells in a range
   */
  mergeCells(
    ws: ExcelJS.Worksheet,
    startRow: number,
    startCol: number,
    endRow: number,
    endCol: number
  ): void {
    ws.mergeCells(startRow, startCol, endRow, endCol);
  }

  /**
   * Set row height
   */
  setRowHeight(ws: ExcelJS.Worksheet, row: number, height: number): void {
    ws.getRow(row).height = height;
  }

  /**
   * Set sheet properties
   */
  setSheetProperties(
    ws: ExcelJS.Worksheet,
    options?: {
      freezePanes?: boolean;
      freezeRow?: number;
      autoFilter?: boolean;
      filterRange?: string;
    }
  ): void {
    if (options?.freezePanes !== false) {
      this.styleManager.freezePanes(ws, options?.freezeRow || 1);
    }

    if (options?.autoFilter && options?.filterRange) {
      ws.autoFilter = options.filterRange;
    }

    // Auto-adjust column widths
    this.styleManager.autoAdjustColumnWidths(ws);
  }

  /**
   * Generate Excel buffer (for saving/downloading)
   */
  async generateBuffer(): Promise<ArrayBuffer> {
    if (!this.workbook) {
      throw new Error('No workbook to save. Call createWorkbook() first.');
    }

    const buffer = await this.workbook.xlsx.writeBuffer();
    return buffer as ArrayBuffer;
  }

  /**
   * Get standard border style
   */
  protected getBorder(): Partial<ExcelJS.Borders> {
    return {
      top: { style: 'thin', color: { argb: 'FF000000' } },
      left: { style: 'thin', color: { argb: 'FF000000' } },
      bottom: { style: 'thin', color: { argb: 'FF000000' } },
      right: { style: 'thin', color: { argb: 'FF000000' } },
    };
  }

  /**
   * Get style manager for custom styling
   */
  getStyleManager(): StyleManager {
    return this.styleManager;
  }
}
